docs/src/references/c8y-configuration-management.md

```diff
@@ -162,4 +164,104 @@ Note that:
+
+The `c8y_configuration_plugin` expects the process/external device always putting the latest config file as retain message to the MQTT topic `{path}`. Then the plugin consumes that retain message whenever a cloud request comes in.
+
+Example Plugin Config:
+```shell
+$ cat /etc/tedge/c8y/c8y-configuration-plugin.toml
+files = [
+    { path = '/etc/tedge/tedge.toml', type = 'tedge.toml' },
+    { path = 'mqtt://configs/bar.conf', type = 'bar.conf', childid = 'child1'  }
+  ]
+```
+
+Example Flow:
+
+**Start Behaviour:**
+  * external device `child1`: starts
+  * external device `child1`: publishs all its config files to MQTT (with retain); Example: Publish to topic `configs/bar.conf`
+
+**Device-to-Cloud Behaviour:**
+  * at some point a config retrieval for type `bar.conf` for `child1` arrives at C8Y config plugin
+  * C8Y config plugin: subscribes to `configs/bar.conf` and gets immediately the config file content from the retained message
+  * C8Y config plugin: sends the recived MQTT payload as config file to C8Y<br/><br/>
+    NOTE: The responsibility to assure that the latest config file content is on the MQTT bus is always on the process/external device.
+When there is no retain message the plugin sends an error to the cloud on an every incoming config retrieval request.
+
+**Cloud-to-Device Behaviour:**
+  * at some point a config sent from cloud for type `bar.conf` for `child1` arrives at C8Y config plugin
+  * C8Y config plugin: publishs the config file content as MQTT retain message to `configs/bar.conf` 
+  * external device `child1`: recognizes the MQTT message on `configs/bar.conf` and processes the new config file content
+
+**Proposal 2:**
+
+The `c8y_configuration_plugin` provides the field `protocol` for all records in the `c8y_configuration_plugin` configuration (reference to section 'Configuration' above). If `protocol` is set to "http" the `c8y_configuration_plugin` expects another process/external device to provide/consume the configuration file via HTTP protocol. Thereby the records field `path` is used with the prefix `http://<thin-edge device IP address>/tedge/configurations/` as URL. Example: `http://192.168.1.6/tedge/configurations/child1/bar.conf` where (192.168.1.6 is the IP address of the thin-edge device)
+
+The `c8y_configuration_plugin` expects the process/external device always putting the latest config file to the thin-edge device. Then the plugin consumes that file whenever a cloud request comes in.<br/> 
+
+Example Configuration:
+```shell
+$ cat /etc/tedge/c8y/c8y-configuration-plugin.toml
+files = [
+    { path = '/etc/tedge/tedge.toml', type = 'tedge.toml' },
+    { path = '/child1/bar.conf', type = 'bar.conf', protocol = 'http', childid = 'child1'  }
+  ]
+```
+
+Example Flow:
```

Comment:
I imagine a different workflow.

* First, files are published over HTTP by the `tedge_agent` and not the `c8y_configuration_plugin`. Because sharing files with child devices and external component is a general need: for configuration files but also software packages and firmware.
* Second, one needs to distinguish files provided by the child devices (current state - or latest known state) from those to be consumed by the child device (desired state). Making the distinction make things clearer, avoid race conditions (where both endpoints update the same file concurrently), and open opportunities for progress tracking (one can check if a device has updated his configuration or not).
* Third, I would use a combination of HTTP and MQTT. HTTP to deliver/upload/download content. MQTT to trigger actions and notify events. Notably any change of a file under the HTTP root is notified over MQTT. MQTT retained messages could be used to deliver+notify in a single request; but I would consider this later.

Hence, here is the workflow I propose:
* `tedge_agent` serves PUT requests on `tedge/current/<child-id>/<any-path-chosen-by the-child-device>`.
* `tedge_agent` serves GET requests on `tedge/desired/<child-id>/<any-path-chosen-by the-cloud>`
* `tedge_agent` notify over MQTT any change of a file under `tedge-http-root` (uploaded by a child device in tedge/current or updated by a local process in tedge/desired).
* `c8y_configuration_plugin` is updated to handle two versions for each configuration (current and desired).
   * On c8y read request for a configuration,  `c8y_configuration_plugin` returns the `current` file content.
   * On c8y write request for a configuration,  `c8y_configuration_plugin` updates the `desired` file content and wait for  the `current` content to be updated to return the operation status to cumulocity. To be clarified: how to deal with errors, the child device behind unable to load a file.
   * The current behavior of `c8y_configuration_plugin` can be simulated with `desired = current = path` when only a `path` is provided (To be checked).
*  To manage configuration files for child-devices from c8y, the configuration of `c8y_configuration_plugin` uses then the HTTP feature of the `tedge_agent`.
   * ` { type = 'bar.conf', child_id = 'child1',  current = '/http-root/current/child1/bar.conf', desired = '/http-root/desired/child1/bar.conf'  }  }`

Reply:
That's an interesting approach. Especially since we would implement that new thin-edge feature (the HTTP GET and PUT service, probably in the tedge_agent), that can be mostly used as is also for other features than C8Y Cfg Plugin. And on the other hand the C8Y Cfg Plugin has just minimal touchpoints with that.

Reply:
I am curious how good that approach applies to Logging Management. One main difference here is, that log-files are no static files, but usually will be appended over their whole lifecycle. To avoid also here a request/response patterm there might be no way than uploading those files on change or in a specific cycle-time to thin-edge. However it would be good to avoid uploading the whole file again and again, and instead allow to upload/append just the new lines.

When I remember right (not completely sure) there was I way with HTTP POST to upload just chunks of a file. To be verified, but that could be another pro for that HTTP proposal.

Reply:
What's about security / access control?

With MQTT there is the ACL that allows to define multiple fine granular access rules, that can be bound to individual certificates or user/password per external device. That is what comes with mosquitto.

We need to provided something similar for HTTP as well. I see two level here:
1) Privilegde on HTTP server level:
To assure just authorised devices can access that HTTP server, and at that level all device have same permissions.
2) Privilegde on HTTP URL/Resource level:
To assure a child-device can only access it's own files, and not the ones of another device.

I am pretty sure we can agree that level 1 is a mandatory.

However I am not sure if level 2 is required. We need stress that and potentially check if we would have any feasible solution available. That depends probably on the server/component we select to use to provide HTTP services.

Reply:
I added that as favored proposal to the spec. See https://github.com/thin-edge/thin-edge.io/pull/1236/files#diff-0d430102a9a7f4e4431390a85806a990b51d2143dbc97bf596a267c893e1f693R194-R265